### PR TITLE
gzip filter can be configured via black-/whitelist in application.conf

### PR DESCRIPTION
**Pull request by mkurz**
(I think) the gzip filter is the only play-provided  filter left which can't be configured solely via `application.conf`.

Would be great if this makes it in before 2.6 final.

### CODE REVIEW

File: documentation/manual/working/commonGuide/filters/GzipEncoding.md [hunk 1]
```diff
@@ -17,7 +17,20 @@ The gzip filter supports a small number of tuning configuration options, which c
 
 ## Controlling which responses are gzipped
 
-To control which responses are and aren't implemented, use the `shouldGzip` parameter, which accepts a function of a request header and a response header to a boolean.
+You can control which responses are and aren't implemented via `application.conf`:
```
Comment (marcospereira) [inline documentation/manual/working/commonGuide/filters/GzipEncoding.md]: It could be good to better explain that this is based on the content type header.

File: documentation/manual/working/commonGuide/filters/GzipEncoding.md [hunk 2]
```diff
@@ -17,7 +17,20 @@ The gzip filter supports a small number of tuning configuration options, which c
...
+
+```
+play.filters.gzip {
+    # If non empty, then a response will only be compressed if its content type is in this list.
+    whiteList = [ "text/html", "text/plain"]
```
Comment (marcospereira) [inline documentation/manual/working/commonGuide/filters/GzipEncoding.md]: I think you can use `whitelist` and `blacklist` (without the camel case).

File: framework/src/play-filters-helpers/src/test/scala/play/filters/gzip/GzipFilterSpec.scala [hunk 3]
```diff
@@ -227,11 +267,13 @@ class GzipFilterSpec extends PlaySpecification with DataTables {
 
   }
 
-  def withApplication[T](result: Result, chunkedThreshold: Int = 1024)(block: Application => T): T = {
+  def withApplication[T](result: Result, chunkedThreshold: Int = 1024, whiteList: List[String] = List(), blackList: List[String] = List())(block: Application => T): T = {
```
Comment (marcospereira) [inline framework/src/play-filters-helpers/src/test/scala/play/filters/gzip/GzipFilterSpec.scala]: Use `List.empty`.

File: documentation/manual/working/commonGuide/filters/GzipEncoding.md (hunk 2 above)
Comment (mkurz) [inline documentation/manual/working/commonGuide/filters/GzipEncoding.md]: I used camel case because it's used in the `reference.conf` everywhere already:
https://github.com/playframework/playframework/blob/master/framework/src/play-filters-helpers/src/main/resources/reference.conf#L74-L82

File: framework/src/play-filters-helpers/src/test/scala/play/filters/gzip/GzipFilterSpec.scala [hunk 4]
```diff
@@ -111,6 +111,46 @@ class GzipFilterSpec extends PlaySpecification with DataTables {
       checkNotGzipped(makeGzipRequest(app), "")(app.materializer)
     }
 
+    "gzip content type which is on the whiteList" in withApplication(Ok("hello").as("text/css"), whiteList = contentTypes) { implicit app =>
```
Comment (marcospereira) [inline framework/src/play-filters-helpers/src/test/scala/play/filters/gzip/GzipFilterSpec.scala]: It could be good to add tests when both whitelist and blacklist are empty.

File: documentation/manual/working/commonGuide/filters/GzipEncoding.md (hunk 2 above)
Comment (marcospereira) [inline documentation/manual/working/commonGuide/filters/GzipEncoding.md]: Good point. Let's keep it consistent then. ;-)

File: framework/src/play-filters-helpers/src/test/scala/play/filters/gzip/GzipFilterSpec.scala (hunk 4 above)
Comment (mkurz) [inline framework/src/play-filters-helpers/src/test/scala/play/filters/gzip/GzipFilterSpec.scala]: For all other tests which have been there before both whitelist and blacklist are empty. So I think this is tested already or not? E.g. check line 45 (`"gzip responses"`) I think this is the test case you are looking for.

**Comment by mkurz** [comment]
@marcospereira Thanks for the review. I addressed your feedback.

File: documentation/manual/working/commonGuide/filters/GzipEncoding.md (hunk 2 above)
Comment (gmethvin) [inline documentation/manual/working/commonGuide/filters/GzipEncoding.md]: I think this should be `contentType.whiteList` and `contentType.blackList`.

File: framework/src/play-filters-helpers/src/main/scala/play/filters/gzip/GzipFilter.scala [hunk 5]
```diff
@@ -207,8 +207,19 @@ object GzipFilterConfig {
...
-    )
+      chunkedThreshold = config.get[ConfigMemorySize]("chunkedThreshold").toBytes.toInt,
+      shouldGzip = (req, res) => {
+      val contentType = res.body.contentType.getOrElse("").toLowerCase;
+
+      val whiteList = config.get[Seq[String]]("whiteList").toSet
+      val blackList = config.get[Seq[String]]("blackList").toSet
+
+      if (whiteList.nonEmpty) {
+        whiteList.exists(contentType.startsWith(_))
```
Comment (gmethvin) [inline framework/src/play-filters-helpers/src/main/scala/play/filters/gzip/GzipFilter.scala]: Would it make sense to use the `MediaRangeParser` and define media ranges instead of using `startsWith`?

**Comment by mkurz** [comment]
@gmethvin I added the `contentType.` prefix. About the `MediaRangeParser` I am not sure right now. I think it was build for something different or not? I think it's ok to stay with `startsWith` for now.

**Comment by gmethvin** [comment]
@mkurz You're right that `MediaRangeParser` does more that what we need here. The relevant part is the `Parser[MediaType]` here: https://github.com/playframework/playframework/blob/master/framework/src/play/src/main/scala/play/api/http/MediaRange.scala#L215. We could factor that out and use it here as well, meaning we can use the same logic for accepts: https://github.com/playframework/playframework/blob/master/framework/src/play/src/main/scala/play/api/http/MediaRange.scala#L52

It seems to make sense to use a common parser for media types and have it work with the same syntax used in HTTP headers.

We should be able to get this into 2.6.0, which will be done in the next few weeks, if not RC2. But if not we can always add it in 2.6.x later.

**Comment by mkurz** [comment]
@gmethvin Can you please check the commit I added? I choose the easiest way for now and just made use of `MediaRange` and `MediaType`.

File: documentation/manual/working/commonGuide/filters/GzipEncoding.md [hunk 6]
```diff
@@ -17,7 +17,24 @@ The gzip filter supports a small number of tuning configuration options, which c
...
-To control which responses are and aren't implemented, use the `shouldGzip` parameter, which accepts a function of a request header and a response header to a boolean.
+You can control which responses are and aren't gzipped based on their content types via `application.conf`:
+
+```
+play.filters.gzip {
+
+    contentType {
+
+        # If non empty, then a response will only be compressed if its content type is in this list.
+        whiteList = [ "text/*", "application/javascript" ]
```
Comment (mkurz) [inline documentation/manual/working/commonGuide/filters/GzipEncoding.md]: Added.

File: framework/src/play-filters-helpers/src/main/scala/play/filters/gzip/GzipFilter.scala [hunk 7]
```diff
@@ -207,8 +207,20 @@ object GzipFilterConfig {
 
     GzipFilterConfig(
       bufferSize = config.get[ConfigMemorySize]("bufferSize").toBytes.toInt,
-      chunkedThreshold = config.get[ConfigMemorySize]("chunkedThreshold").toBytes.toInt
-    )
+      chunkedThreshold = config.get[ConfigMemorySize]("chunkedThreshold").toBytes.toInt,
+      shouldGzip = (req, res) => {
+      res.body.contentType.flatMap(ct => MediaType.parse(ct.toLowerCase)).map(mt => mt.mediaType + "/" + mt.mediaSubType).map(mimeType => {
```
Comment (mkurz) [inline framework/src/play-filters-helpers/src/main/scala/play/filters/gzip/GzipFilter.scala]: Done. However your example code isn't working 1:1. E.g `MediaType.parse(...)` returns an Option.

Please review again, thanks!

Comment (gmethvin) [inline framework/src/play-filters-helpers/src/main/scala/play/filters/gzip/GzipFilter.scala]: You're right. I didn't compile the code. I think you want to default the `mimeType` string to whatever the content type was in the `None` case.

Comment (mkurz) [inline framework/src/play-filters-helpers/src/main/scala/play/filters/gzip/GzipFilter.scala]: See my last commit, is it OK like this?

File: framework/src/play-filters-helpers/src/main/scala/play/filters/gzip/GzipFilter.scala [hunk 8]
```diff
@@ -204,11 +204,25 @@ object GzipFilterConfig {
...
+    val whiteList = MediaRange.parse(config.get[Seq[String]]("contentType.whiteList").mkString(", "))
+    val blackList = MediaRange.parse(config.get[Seq[String]]("contentType.blackList").mkString(", "))
 
     GzipFilterConfig(
       bufferSize = config.get[ConfigMemorySize]("bufferSize").toBytes.toInt,
-      chunkedThreshold = config.get[ConfigMemorySize]("chunkedThreshold").toBytes.toInt
-    )
+      chunkedThreshold = config.get[ConfigMemorySize]("chunkedThreshold").toBytes.toInt,
+      shouldGzip = (req, res) => {
+      res.body.contentType match {
```
Comment (gmethvin) [inline framework/src/play-filters-helpers/src/main/scala/play/filters/gzip/GzipFilter.scala]: Now that I notice `MediaType.parse` has `unapply` I think the best way to do this is:

```scala
res.body.contentType match {
  case Some(MediaType.parse(mt)) =>
    val mimeType = s"${mt.mediaType}/${mt.mediaSubType}"
    if (whiteList.nonEmpty) {
      whiteList.exists(_.accepts(mimeType))
    } else {
      blacklist.forall(!_.accepts(mimeType))
    }
  case _ => whiteList.isEmpty
}
```
We know an invalid content type will not match anything in either the blacklist or whitelist. So the failure case for that should also be `whiteList.isEmpty`, just like having no content type.

Comment (mkurz) [inline framework/src/play-filters-helpers/src/main/scala/play/filters/gzip/GzipFilter.scala]: @gmethvin Updated. Please review once more, thanks!
